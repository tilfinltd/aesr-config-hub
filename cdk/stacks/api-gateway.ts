import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  DomainName,
  RestApi,
  CognitoUserPoolsAuthorizer,
  MockIntegration,
  PassthroughBehavior,
  Model,
  LambdaIntegration,
} from "aws-cdk-lib/aws-apigateway";
import { ICertificate } from "aws-cdk-lib/aws-certificatemanager";
import { IUserPool } from "aws-cdk-lib/aws-cognito";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { ARecord, HostedZone, RecordTarget } from "aws-cdk-lib/aws-route53";
import { ApiGatewayDomain } from "aws-cdk-lib/aws-route53-targets";

export interface ApiGatewayStackProps extends StackProps {
  prefix: string;
  subDomain: string;
  certificate: ICertificate;
  userPool: IUserPool;
  configProviderLambdaName: string;
}

export class ApiGatewayStack extends Stack {
  constructor(scope: Construct, id: string, props: ApiGatewayStackProps) {
    super(scope, id, props);
    const { prefix, subDomain, certificate } = props;

    const rootDomain = `${subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    const apiDomain = `api.${subDomain}.aesr.dev`;
    const restApi = new RestApi(this, `${prefix}-api`);

    const domainName = new DomainName(this, "ApiDomain", {
      domainName: apiDomain,
      certificate,
    });

    domainName.addBasePathMapping(restApi);

    new ARecord(this, "ApiDomainRecord", {
      zone: hostedZone,
      recordName: apiDomain,
      target: RecordTarget.fromAlias(new ApiGatewayDomain(domainName)),
    });

    const authorizer = new CognitoUserPoolsAuthorizer(
      this,
      "CognitoAuthorizer",
      {
        authorizerName: "CognitoAuthorizer",
        cognitoUserPools: [props.userPool],
      },
    );

    const callbackRes = restApi.root.addResource("callback");
    callbackRes.addMethod(
      "GET",
      new MockIntegration({
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
        requestTemplates: {
          "application/json": '{"statusCode":200}',
        },
        integrationResponses: [
          {
            statusCode: "200",
            responseTemplates: {
              "text/html": `<html>
            <body style="font-family:sans-serif;padding:36px">
             <p style="font-size:18px;line-height:2">
               Please keep this tab open and<br>
               <u>click on the <strong>AWS Extend Switch Roles</strong> icon</u>
               in your browser toolbar<br>
               to complete the connection to <strong>AESR Config Hub</strong>.
             </p>
            </body>
           </html>`,
            },
          },
        ],
      }),
      {
        methodResponses: [
          {
            statusCode: "200",
            responseModels: { "text/html": Model.EMPTY_MODEL },
          },
        ],
      },
    );

    const configProviderLambda = lambda.Function.fromFunctionName(
      this,
      "ConfigProviderLambda",
      props.configProviderLambdaName,
    );
    const userRes = restApi.root.addResource("user");
    const configRes = userRes.addResource("config");
    configRes.addMethod("GET", new LambdaIntegration(configProviderLambda), {
      authorizer,
    });
  }
}
