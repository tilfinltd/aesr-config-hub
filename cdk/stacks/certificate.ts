import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  Certificate,
  CertificateValidation,
  ICertificate,
} from "aws-cdk-lib/aws-certificatemanager";
import { HostedZone } from "aws-cdk-lib/aws-route53";

export interface CertificateStackProps extends StackProps {
  subDomain: string;
}

export class GlobalCertificateStack extends Stack {
  public readonly authCertificate: ICertificate;

  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);

    const rootDomain = `${props.subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    this.authCertificate = new Certificate(this, "AuthDomainCertificate", {
      domainName: `auth.${rootDomain}`,
      validation: CertificateValidation.fromDns(hostedZone),
    });
  }
}

export class LocalCertificateStack extends Stack {
  public readonly apiCertificate: ICertificate;

  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);

    const rootDomain = `${props.subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    this.apiCertificate = new Certificate(this, "ApiDomainCertificate", {
      domainName: `api.${rootDomain}`,
      validation: CertificateValidation.fromDns(hostedZone),
    });
  }
}
