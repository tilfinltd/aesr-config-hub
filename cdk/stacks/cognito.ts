import { Duration, RemovalPolicy, Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { ICertificate } from "aws-cdk-lib/aws-certificatemanager";
import { IUserPool, OAuthScope, UserPool } from "aws-cdk-lib/aws-cognito";
import { ARecord, HostedZone, RecordTarget } from "aws-cdk-lib/aws-route53";
import { UserPoolDomainTarget } from "aws-cdk-lib/aws-route53-targets";

export interface CognitoStackProps extends StackProps {
  prefix: string;
  subDomain: string;
  certificate: ICertificate;
  refreshTokenValidity: Duration;
}

export class CognitoStack extends Stack {
  public readonly userPool: IUserPool;

  constructor(scope: Construct, id: string, props: CognitoStackProps) {
    super(scope, id, props);
    const { prefix, subDomain, certificate } = props;

    const rootDomain = `${subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    // For Cognito custom domain, an A record for the root domain must exist.
    new ARecord(this, "RootDomainRecord", {
      zone: hostedZone,
      recordName: rootDomain,
      target: RecordTarget.fromIpAddresses("127.0.0.1"),
    });

    const authDomain = `auth.${subDomain}.aesr.dev`;
    const callbackUrl = `https://api.${subDomain}.aesr.dev/callback`;

    const userPool = new UserPool(this, "UserPool", {
      userPoolName: `${prefix}-user-pool`,
      standardAttributes: {
        email: { required: true, mutable: true },
      },
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const customDomain = userPool.addDomain("UserPoolDomain", {
      customDomain: {
        domainName: authDomain,
        certificate,
      },
    });

    new ARecord(this, "AuthDomainRecord", {
      zone: hostedZone,
      recordName: authDomain,
      target: RecordTarget.fromAlias(new UserPoolDomainTarget(customDomain)),
    });

    userPool.addClient("UserPoolClient", {
      userPoolClientName: `${prefix}-browser-extension`,
      generateSecret: false,
      oAuth: {
        flows: {
          authorizationCodeGrant: true,
        },
        callbackUrls: [callbackUrl],
        scopes: [OAuthScope.OPENID, OAuthScope.EMAIL],
      },
      idTokenValidity: Duration.hours(1),
      refreshTokenValidity: props.refreshTokenValidity,
    });

    this.userPool = userPool;
  }
}
