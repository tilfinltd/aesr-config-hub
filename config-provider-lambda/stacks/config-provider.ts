import { Duration, Stack, StackProps, aws_lambda_nodejs } from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  Effect,
  ManagedPolicy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { Bucket } from "aws-cdk-lib/aws-s3";

export interface ConfigProviderStackProps extends StackProps {
  userConfigBucketName: string;
  lambdaName: string;
}

export class ConfigProviderStack extends Stack {
  constructor(scope: Construct, id: string, props: ConfigProviderStackProps) {
    super(scope, id, props);
    const { userConfigBucketName, lambdaName } = props;

    const lambdaRole = new Role(this, "ConfigProviderLambdaExecutionRole", {
      roleName: `${lambdaName}-LambdaExecutionRole`,
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole",
        ),
      ],
    });

    new aws_lambda_nodejs.NodejsFunction(this, "ConfigProviderLambda", {
      functionName: lambdaName,
      runtime: lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: [
          '@aws-sdk/*', // Use the AWS SDK for JS v3 available in the Lambda runtime
        ],
      },
      entry: "lambda/index.ts",
      handler: "handler",
      memorySize: 128,
      timeout: Duration.seconds(10),
      role: lambdaRole,
      environment: {
        USER_CONFIG_BUCKET: userConfigBucketName,
      },
    });

    const logGroup = new LogGroup(this, "ConfigProviderLogGroup", {
      logGroupName: `/aws/lambda/${lambdaName}`,
    });
    logGroup.grantWrite(lambdaRole);

    const userConfigBucket = new Bucket(this, "UserConfigBucket", {
      bucketName: userConfigBucketName,
    });

    userConfigBucket.addToResourcePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ["s3:GetObject"],
        principals: [lambdaRole],
        resources: [userConfigBucket.bucketArn + "/*"],
      }),
    );
  }
}
