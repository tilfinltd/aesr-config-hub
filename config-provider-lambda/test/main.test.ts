import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ConfigProviderStack } from "../stacks/config-provider";

const cdkJson = require("../cdk.json");

describe("All stacks", () => {
  test("default", () => {
    const app = new App(cdkJson);
    const env = {
      region: "ap-northeast-1",
      account: "000011112222",
    };
    const prefix = "aesr-test";

    const configProviderLambdaStack = new ConfigProviderStack(
      app,
      `${prefix}-config-provider`,
      {
        bucketName: `${prefix}-config-provider-${env.account}`,
        lambdaName: `${prefix}-config-proivder`,
      },
    );
    const configProviderTemplate = Template.fromStack(
      configProviderLambdaStack,
    ).toJSON();
    expect(configProviderTemplate).toMatchSnapshot();
  });
});
