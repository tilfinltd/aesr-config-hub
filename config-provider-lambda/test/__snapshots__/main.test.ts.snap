// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All stacks default 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AesrConfigBucket62E33CAE": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "aesr-test-config-provider-000011112222",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "AesrConfigBucketPolicy8A028DBC": {
      "Properties": {
        "Bucket": {
          "Ref": "AesrConfigBucket62E33CAE",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "ConfigProviderLambdaExecutionRole707DC047",
                    "Arn",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "AesrConfigBucket62E33CAE",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ConfigProviderLambda3FAD5AEB": {
      "DependsOn": [
        "ConfigProviderLambdaExecutionRoleDefaultPolicy8DD8598B",
        "ConfigProviderLambdaExecutionRole707DC047",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "463d3a7b236c2382c30cc8f383c66f9934e8c938c4df69a2748f3b291854789b.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CONFIG_BUCKET_NAME": "aesr-test-config-provider-000011112222",
          },
        },
        "FunctionName": "aesr-test-config-proivder",
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "ConfigProviderLambdaExecutionRole707DC047",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConfigProviderLambdaExecutionRole707DC047": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "aesr-test-config-proivder-LambdaExecutionRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "ConfigProviderLambdaExecutionRoleDefaultPolicy8DD8598B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ConfigProviderLogGroup298AC8F1",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ConfigProviderLambdaExecutionRoleDefaultPolicy8DD8598B",
        "Roles": [
          {
            "Ref": "ConfigProviderLambdaExecutionRole707DC047",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ConfigProviderLogGroup298AC8F1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "/aws/lambda/aesr-test-config-proivder",
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
